# Segmenting and Clustering Neighborhoods in Toronto

In this notebook, I will explore and cluster neighborhoods in the city of Toronto, Canada. The data I will be used come from Wikipedia page provided in the following link : https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M. I will scrape that page in order to obtain the data in the table of postal codes and transform the data into a pandas dataframe.

# Import All Libraries and Packages Needed

In [1]:
import pandas as pd
import csv
import requests
from bs4 import BeautifulSoup
from collections import OrderedDict
import numpy as np
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Web Scraping

For scraping, I will use pandas and BeautifulSoup package. All of libraries and packages needed are installed and imported in cell above. 

First, get the URL of the website 

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [3]:
soup = BeautifulSoup(website_url, 'html.parser')

Since I will scrape the data from the table, so let's find the table with keyword 'table' from HTML and define it into a new variable called 'my_table'.

In [4]:
my_table = soup.find('table', {'class' : 'wikitable sortable'})

#### Convert the data in the table to CSV file named 'toronto.csv'.

In [5]:
with open('toronto.csv', 'w',newline='') as csvfile:
    f = csv.writer(csvfile)
    f.writerow(['PostCode', 'Borough', 'Neighborhood'])  # headers
    for x in my_table:
        rows = my_table.find_all('tr') #find all tr tag
        for tr in rows:
            data=[]
            cols = tr.find_all('td') #find all td tags
            for td in cols:
                data.append(td.text.strip()) 
            f.writerow(data)
    
    print('toronto.csv was created!')

toronto.csv was created!


So far, the new CSV file created and set to be used for exploring. The next step is to load the dataset into pandas dataframe and to explore the dataset. 

# Load and Explore Dataset

Load the dataset into pandas dataframe.

In [6]:
df = pd.read_csv('toronto.csv')
df.head()

PostCode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [7]:
df.shape

(574, 3)

The dataset has 574 rows and 3 columns. As we can see, in columns Borough and Neighborhood still have some values of Not assigned. I will process the rows that have an assigned Borough, so let's remove the rows that have a Not assigned Borough.

In [8]:
df = df.drop(df[df.Borough == 'Not assigned'].index)

Let's take a look at the first 10 rows after removing.

In [9]:
df.head(10)

PostCode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A      Queen's Park      Not assigned
9       M9A      Queen's Park      Queen's Park
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

Let's take a look at unique values of the Borough.

In [10]:
df['Borough'].unique()

array(['North York', 'Downtown Toronto', "Queen's Park", 'Scarborough',
       'East York', 'Etobicoke', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

Now, let's check the dimension of the table after removing.

In [11]:
df.shape

(420, 3)

As we can see, now we have 420 rows and 3 columns in the table after removing some an Not assigned boroughs.

After dealing with Not assigned values in Borough column, now let's take care of some Not assigned values in Neighborhood column. For Not assigned values in Neighborhood column, unlike in Borough column, I will replace Not assined values with value of its borough. So, it will contain same value as its borough.

In [12]:
df['Neighborhood'] = df['Neighborhood'].replace('Not assigned', df['Borough'])
df.head(10)

PostCode           Borough      Neighborhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M6A        North York  Lawrence Heights
6       M6A        North York    Lawrence Manor
7       M7A      Queen's Park      Queen's Park
9       M9A      Queen's Park      Queen's Park
10      M1B       Scarborough             Rouge
11      M1B       Scarborough           Malvern
13      M3B        North York   Don Mills North

Now, I will divide the data into two new dataframes for making easier to process. 

First, I will create a new variable that contain the dataframe of postal codes and neighborhoods named 'df_pn'. I will join the neighborhoods that have same postal code.

In [13]:
df_pn = df.groupby('PostCode')['Neighborhood'].apply(lambda tags: ', '.join(tags)).reset_index()

df_pn['Neighborhood'] = (df_pn['Neighborhood'].str.split()
                              .apply(lambda x: OrderedDict.fromkeys(x).keys())
                              .str.join(' ')
                              .str.rsplit(', ',1).str[0])

df_pn.head(10)

PostCode                                     Neighborhood
0      M1B                                   Rouge, Malvern
1      M1C           Highland Creek, Rouge Hill, Port Union
2      M1E                Guildwood, Morningside, West Hill
3      M1G                                           Woburn
4      M1H                                        Cedarbrae
5      M1J                              Scarborough Village
6      M1K                    East Birchmount Park, Ionview
7      M1L                  Clairlea, Golden Mile, Oakridge
8      M1M  Cliffcrest, Cliffside, Scarborough Village West
9      M1N                      Birch Cliff, Cliffside West

As we can see, now the neighborhoods have merged based on its postal codes.

Now, let's create another new variable named 'df_pb' that contain the dataframe of postal codes and boroughs. I will remove the duplicated data, so that the dataframe only contain unique values for PostCode, like previous dataframe.

In [14]:
df_pb = df.drop(['Neighborhood'], axis=1)
df_pb = df_pb.drop_duplicates()
df_pb.head(20)

PostCode           Borough
2       M3A        North York
3       M4A        North York
4       M5A  Downtown Toronto
5       M6A        North York
7       M7A      Queen's Park
9       M9A      Queen's Park
10      M1B       Scarborough
13      M3B        North York
14      M4B         East York
16      M5B  Downtown Toronto
18      M6B        North York
21      M9B         Etobicoke
26      M1C       Scarborough
30      M3C        North York
32      M4C         East York
33      M5C  Downtown Toronto
34      M6C              York
37      M9C         Etobicoke
41      M1E       Scarborough
46      M4E      East Toronto

Now, let's concatenate the two dataframes based on PostCode.

In [15]:
df_fix = pd.concat([df_pb.set_index('PostCode'),df_pn.set_index('PostCode')], axis=1, join='inner').reset_index()
df_fix.head(10)

PostCode           Borough                     Neighborhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto                     Harbourfront
3      M6A        North York          Lawrence Heights, Manor
4      M7A      Queen's Park                     Queen's Park
5      M9A      Queen's Park                     Queen's Park
6      M1B       Scarborough                   Rouge, Malvern
7      M3B        North York                  Don Mills North
8      M4B         East York  Woodbine Gardens, Parkview Hill
9      M5B  Downtown Toronto         Ryerson, Garden District

let's take a look the dimension of the dataframe.

In [16]:
df_fix.shape

(103, 3)

# Get Latitude and Longitude Coordinates from different CSV File

In order to utilize the Foursquare location data, I will add the latitude and longitude coordinates of each neighborhood from different CSV file provided.

Now, load the geospatial coordinates of Toronto data into pandas dataframe.

In [17]:
df_coor = pd.read_csv('Geospatial_Coordinates.csv')
df_coor.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Then, merge the dataframe with the previous dataframe based on postal codes.

In [18]:
df_fix = pd.concat([df_fix.set_index('PostCode'),df_coor.set_index('Postal Code')], axis=1, join='inner').reset_index()
df_fix.head(10)

index           Borough                     Neighborhood   Latitude  \
0   M3A        North York                        Parkwoods  43.753259   
1   M4A        North York                 Victoria Village  43.725882   
2   M5A  Downtown Toronto                     Harbourfront  43.654260   
3   M6A        North York          Lawrence Heights, Manor  43.718518   
4   M7A      Queen's Park                     Queen's Park  43.662301   
5   M9A      Queen's Park                     Queen's Park  43.667856   
6   M1B       Scarborough                   Rouge, Malvern  43.806686   
7   M3B        North York                  Don Mills North  43.745906   
8   M4B         East York  Woodbine Gardens, Parkview Hill  43.706397   
9   M5B  Downtown Toronto         Ryerson, Garden District  43.657162   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494  
5 -79.532242  
6 -79.194353  
7 -79.352188  
8 -79.309937  
9 -79.378937

As we can see, the postal code column still named 'index', so I will rename that column to be 'PostCode'.

In [19]:
df_fix = df_fix.rename(columns={'index' : 'PostCode'})
df_fix.head()

PostCode           Borough             Neighborhood   Latitude  Longitude
0      M3A        North York                Parkwoods  43.753259 -79.329656
1      M4A        North York         Victoria Village  43.725882 -79.315572
2      M5A  Downtown Toronto             Harbourfront  43.654260 -79.360636
3      M6A        North York  Lawrence Heights, Manor  43.718518 -79.464763
4      M7A      Queen's Park             Queen's Park  43.662301 -79.389494

Let's count the boroughs and neighborhoods in the dataframe.

In [20]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_fix['Borough'].unique()),
        df_fix.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto.

In [21]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto with neighborhoods.

In [22]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_fix['Latitude'], df_fix['Longitude'], df_fix['Borough'], df_fix['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### In case generated map does not appear on github page, so I took a screenshoot of the map

<img src = "image-toronto\map-toronto.JPG" width = 800 /> ![](image-toronto\map-toronto.jpg)

Let's simplify the above map as well as segment and cluster only the neighborhoods in Downtown Toronto. So, let's create a new dataframe of Downtown Toronto data named 'df_dtoronto'.

In [23]:
df_dtoronto = df_fix[df_fix['Borough']=='Downtown Toronto'].reset_index(drop=True)
df_dtoronto.head()

PostCode           Borough              Neighborhood   Latitude  Longitude
0      M5A  Downtown Toronto              Harbourfront  43.654260 -79.360636
1      M5B  Downtown Toronto  Ryerson, Garden District  43.657162 -79.378937
2      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418
3      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306
4      M5G  Downtown Toronto        Central Bay Street  43.657952 -79.387383

Let's take a look at the dimension of the dataframe. Make sure that the dataframe has 18 rows and 5 columns.

In [24]:
df_dtoronto.shape

(18, 5)

Now, let's get the geographical coordinates of Downtown Toronto.

In [25]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


Then, as we did with Toronto, let's visualize Downtown Toronto with the neighborhoods in it.

In [26]:
# create map of Downtown Toronto using latitude and longitude values
map_dtoronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_dtoronto['Latitude'], df_dtoronto['Longitude'], df_dtoronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_dtoronto)  
    
map_dtoronto

### In case generated map does not appear on github page, so I took a screenshoot of the map

<img src = "image-toronto\map-dtoronto.jpg" width = 800 /> ![](image-toronto\map-dtoronto.jpg)

Next, I am going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

# Define Foursquare Credentials and Version

In [27]:
CLIENT_ID = 'QGPLMTPPGNBAGNTFZQ1WRL3OABORRQ52O5SMGTXMFJZUOUHR' # your Foursquare ID
CLIENT_SECRET = 'HRQFV3VDCZTYIKMP0GCSRBW3MMDBN5LCGXB0FM4WRWLWBSYJ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QGPLMTPPGNBAGNTFZQ1WRL3OABORRQ52O5SMGTXMFJZUOUHR
CLIENT_SECRET:HRQFV3VDCZTYIKMP0GCSRBW3MMDBN5LCGXB0FM4WRWLWBSYJ


#### Let's explore the second neighborhood in our dataframe.

In [28]:
df_dtoronto.loc[1, 'Neighborhood']

'Ryerson, Garden District'

Get the neighborhood's latitude and longitude values.

In [29]:
neighborhood_latitude = df_dtoronto.loc[1, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_dtoronto.loc[1, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_dtoronto.loc[1, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Ryerson, Garden District are 43.6571618, -79.37893709999999.


#### Now, let's get the top 200 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [30]:
LIMIT = 200 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=QGPLMTPPGNBAGNTFZQ1WRL3OABORRQ52O5SMGTXMFJZUOUHR&client_secret=HRQFV3VDCZTYIKMP0GCSRBW3MMDBN5LCGXB0FM4WRWLWBSYJ&v=20180605&ll=43.6571618,-79.37893709999999&radius=500&limit=200'

Send the GET request and examine the resutls

In [31]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5de36169ad1ab4001e1200b0'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Downtown Toronto',
  'headerFullLocation': 'Downtown Toronto, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 127,
  'suggestedBounds': {'ne': {'lat': 43.6616618045, 'lng': -79.37272880013877},
   'sw': {'lat': 43.652661795499995, 'lng': -79.3851453998612}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '57eda381498ebe0e6ef40972',
       'name': 'UNIQLO ユニクロ',
       'location': {'address': '220 Yonge St',
        'crossStreet': 'at Dundas St W',
        'lat': 43.65591027779457,
        'lng': -79.38064099181345,
        'labeledLatLngs': [{'

Create a function that extracts the category of the venue

In [32]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [33]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name      categories        lat        lng
0          UNIQLO ユニクロ  Clothing Store  43.655910 -79.380641
1  Silver Snail Comics      Comic Shop  43.657031 -79.381403
2          Blaze Pizza     Pizza Place  43.656518 -79.380015
3  Yonge-Dundas Square           Plaza  43.656054 -79.380495
4        Page One Cafe            Café  43.657772 -79.376073

Count the venues were returned by Foursquare

In [34]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues were returned by Foursquare.


# Explore Neighborhoods in Downtown Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Downtown Toronto

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### create a new dataframe called 'dtoronto_venues'.

In [36]:
dtoronto_venues = getNearbyVenues(names=df_dtoronto['Neighborhood'],
                                   latitudes=df_dtoronto['Latitude'],
                                   longitudes=df_dtoronto['Longitude']
                                  )

Harbourfront
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
First Canadian Place, Underground city
Church and Wellesley


Now, let's check the size of the resulting dataframe

In [37]:
print(dtoronto_venues.shape)
dtoronto_venues.head()

(1281, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront               43.65426              -79.360636   
1  Harbourfront               43.65426              -79.360636   
2  Harbourfront               43.65426              -79.360636   
3  Harbourfront               43.65426              -79.360636   
4  Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653191       -79.357947   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

         Venue Category  
0                Bakery  
1           Coffee Shop  
2  Gym / Fitness Center  
3                   Spa  
4        Breakfast Spot

Let's check how many venues were returned for each neighborhood

In [38]:
dtoronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            58   
CN Tower, Bathurst Quay, Island airport, Harbou...                     16   
Cabbagetown, St. James Town                                            47   
Central Bay Street                                                     80   
Chinatown, Grange Park, Kensington Market                              92   
Christie                                                               17   
Church and Wellesley                                                   84   
Commerce Court, Victoria Hotel                                        100   
Design Exchange, Toronto Dominion Centre                              100   
First Canadian Place, Underground city                                100   
Harbord, University of Toronto                                         37   
Harbourfront                                                           48   
Harbourfront East, Toronto Islands, Union Station                     100   
Rosedale                                                                4   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             58   
CN Tower, Bathurst Quay, Island airport, Harbou...                      16   
Cabbagetown, St. James Town                                             47   
Central Bay Street                                                      80   
Chinatown, Grange Park, Kensington Market                               92   
Christie                                                                17   
Church and Wellesley                                                    84   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
First Canadian Place, Underground city                                 100   
Harbord, University of Toronto                                          37   
Harbourfront                                                            48   
Harbourfront East, Toronto Islands, Union Station                      100   
Rosedale                                                                 4   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         98   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Berczy Park                                            58              58   
CN Tower, Bathurst Quay, Island airport, Harbou...     16              16   
Cabbagetown, St. James Town                            47              47   
Central Bay Street                                     80              80   
Chinatown, Grange Park, Kensington Market              92              92   
Christie                                               17              17   
Church and Wellesley                                   84              84   
Commerce Court, Victoria Hotel                        100             100   
Design Exchange, Toronto Dominion Centre              100             100   
First Canadian Place, Unde

Let's find out how many unique categories can be curated from all the returned venues

In [39]:
print('There are {} uniques categories.'.format(len(dtoronto_venues['Venue Category'].unique())))

There are 204 uniques categories.


# Analyze Each Neighborhood

In [40]:
# one hot encoding
dtoronto_onehot = pd.get_dummies(dtoronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
dtoronto_onehot['Neighborhood'] = dtoronto_venues['Neighborhood'] 

dtoronto_onehot.head()

Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               1        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0             0              0        0     0            0   
1             0              0        0     0            0   
2             0              0        0     0            0   
3             0              0        0     0            0   
4             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0            0                   0               0   

   Church  Clothing Store  Cocktail Bar 

In [41]:
dtoronto_onehot.shape

(1281, 204)

Let's move neighborhood column to the first column.

In [42]:
# move neighborhood column to the first column

fixed_columns = [dtoronto_onehot.columns[138]] + list(dtoronto_onehot.columns[0:138]) + list(dtoronto_onehot.columns[139:])
dtoronto_onehot = dtoronto_onehot[fixed_columns]

dtoronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0  Harbourfront                  0        0                   0             0   
1  Harbourfront                  0        0                   0             0   
2  Harbourfront                  0        0                   0             0   
3  Harbourfront                  0        0                   0             0   
4  Harbourfront                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               1        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Candy Store  \
0             0              0        0     0            0   
1             0              0        0     0            0   
2             0              0        0     0            0   
3             0              0        0     0            0   
4             0              0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0          

In [43]:
dtoronto_onehot.shape

(1281, 204)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [44]:
dtoronto_grouped = dtoronto_onehot.groupby('Neighborhood').mean().reset_index()
dtoronto_grouped

Neighborhood  Afghan Restaurant  \
0                            Adelaide, King, Richmond           0.000000   
1                                         Berczy Park           0.000000   
2   CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
3                         Cabbagetown, St. James Town           0.000000   
4                                  Central Bay Street           0.000000   
5           Chinatown, Grange Park, Kensington Market           0.000000   
6                                            Christie           0.000000   
7                                Church and Wellesley           0.011905   
8                      Commerce Court, Victoria Hotel           0.000000   
9            Design Exchange, Toronto Dominion Centre           0.000000   
10             First Canadian Place, Underground city           0.000000   
11                     Harbord, University of Toronto           0.000000   
12                                       Harbourfront           0.000000   
13  Harbourfront East, Toronto Islands, Union Station           0.000000   
14                                           Rosedale           0.000000   
15                           Ryerson, Garden District           0.000000   
16                                     St. James Town           0.000000   
17                    Stn A PO Boxes 25 The Esplanade           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0625              0.0625        0.0625           0.125   
3    0.0000              0.0000        0.0000           0.000   
4    0.0000              0.0000        0.0000           0.000   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000              0.0000        0.0000           0.000   
15   0.0000              0.0000        0.0000           0.000   
16   0.0000              0.0000        0.0000           0.000   
17   0.0000              0.0000        0.0000           0.000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0            0.0000            0.0000             0.030000      0.000000   
1            0.0000            0.0000             0.000000      0.000000   
2            0.1875            0.0625             0.000000      0.000000   
3            0.0000            0.0000             0.000000      0.000000   
4            0.0000            0.0000             0.012500      0.000000   
5            0.0000            0.0000             0.000000      0.000000   
6            0.0000            0.0000             0.000000      0.000000   
7            0.0000            0.0000             0.011905      0.000000   
8            0.0000            0.0000             0.040000      0.000000   
9            0.0000            0.0000             0.040000      0.000000   
10           0.0000            0.0000             0.030000      0.000000   
11           0.0000            0.0000             0.000000      0.000000   
12           0.0000            0.0000             0.000000      0.020833   
13           0.0000            0.0000             0.000000      0.000000   
14           0.0000            0.0000             0.000000      0.000000   
15           0.0000            0.0000             0.010000      0.000000   
16           0.0000            0.0000             0.020000      0.00

In [45]:
dtoronto_grouped.shape

(18, 204)

#### Let's print each neighborhood along with the top 5 most common venues

In [46]:
num_top_venues = 5

for hood in dtoronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = dtoronto_grouped[dtoronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2  Thai Restaurant  0.04
3              Bar  0.04
4       Steakhouse  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1          Bakery  0.05
2  Farmers Market  0.03
3            Café  0.03
4        Beer Bar  0.03


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
             venue  freq
0  Airport Service  0.19
1   Airport Lounge  0.12
2         Boutique  0.06
3  Harbor / Marina  0.06
4      Coffee Shop  0.06


----Cabbagetown, St. James Town----
         venue  freq
0  Coffee Shop  0.09
1   Restaurant  0.06
2          Pub  0.04
3         Park  0.04
4       Bakery  0.04


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.15
1  Italian Restaurant  0.05
2                Café  0.04
3      Ice Cream Shop  0.04
4      Sandwich Place  0.04


----Chinatown, Grange

#### Let's put that into a *pandas* dataframe

First, write a function to sort the venues in descending order.

In [47]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Then, create the new dataframe and display the top 10 venues for each neighborhood.

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = dtoronto_grouped['Neighborhood']

for ind in np.arange(dtoronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dtoronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   
3                        Cabbagetown, St. James Town           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar       Thai Restaurant   
1                Bakery          Cocktail Bar    Seafood Restaurant   
2        Airport Lounge         Boat or Ferry       Harbor / Marina   
3            Restaurant                Market                  Park   
4    Italian Restaurant                  Café          Burger Joint   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                Bakery      Asian Restaurant   
1        Farmers Market              Beer Bar                  Café   
2              Boutique   Rental Car Location                   Bar   
3    Italian Restaurant           Pizza Place                Bakery   
4        Sandwich Place        Ice Cream Shop                   Spa   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant          Burger Joint    American Restaurant  
1           Cheese Shop            Steakhouse         Breakfast Spot  
2           Coffee Shop      Sculpture Garden     Airport Food Court  
3    Chinese Restaurant                   Pub                   Café  
4                   Bar    Chinese Restaurant    Japanese Restaurant

# Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [49]:
# set number of clusters
kclusters = 5

dtoronto_grouped_clustering = dtoronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dtoronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 0, 3, 4, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [50]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

dtoronto_merged = df_dtoronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
dtoronto_merged = dtoronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

dtoronto_merged.head() # check the last columns!

PostCode           Borough              Neighborhood   Latitude  Longitude  \
0      M5A  Downtown Toronto              Harbourfront  43.654260 -79.360636   
1      M5B  Downtown Toronto  Ryerson, Garden District  43.657162 -79.378937   
2      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418   
3      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306   
4      M5G  Downtown Toronto        Central Bay Street  43.657952 -79.387383   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                  Park   
1               0           Coffee Shop        Clothing Store   
2               0                  Café           Coffee Shop   
3               0           Coffee Shop                Bakery   
4               0           Coffee Shop    Italian Restaurant   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                     Bakery                   Pub               Theater   
1  Middle Eastern Restaurant        Cosmetics Shop  Fast Food Restaurant   
2                 Restaurant    Italian Restaurant                 Hotel   
3               Cocktail Bar    Seafood Restaurant        Farmers Market   
4                       Café          Burger Joint        Sandwich Place   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Mexican Restaurant                  Café            Restaurant   
1                  Café                Bakery   Japanese Restaurant   
2        Clothing Store                Bakery        Cosmetics Shop   
3              Beer Bar                  Café           Cheese Shop   
4        Ice Cream Shop                   Spa                   Bar   

  9th Most Common Venue 10th Most Common Venue  
0        Breakfast Spot                Brewery  
1       Bubble Tea Shop    Sporting Goods Shop  
2                 Diner               Beer Bar  
3            Steakhouse         Breakfast Spot  
4    Chinese Restaurant    Japanese Restaurant

Finally, let's visualize the resulting clusters

In [51]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dtoronto_merged['Latitude'], dtoronto_merged['Longitude'], dtoronto_merged['Neighborhood'], dtoronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### In case generated map does not appear on github page, so I took a screenshoot of the map

<img src = "image-toronto\map-cluster-dtoronto.jpg" width = 800 /> ![](image-toronto\map-cluster-dtoronto.jpg)

# Examine Clusters

### Cluster 1

In [52]:
dtoronto_merged.loc[dtoronto_merged['Cluster Labels'] == 0, dtoronto_merged.columns[[2] + list(range(5, dtoronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                        Harbourfront               0   
1                            Ryerson, Garden District               0   
2                                      St. James Town               0   
3                                         Berczy Park               0   
4                                  Central Bay Street               0   
6                            Adelaide, King, Richmond               0   
7   Harbourfront East, Toronto Islands, Union Station               0   
8            Design Exchange, Toronto Dominion Centre               0   
9                      Commerce Court, Victoria Hotel               0   
14                    Stn A PO Boxes 25 The Esplanade               0   
15                        Cabbagetown, St. James Town               0   
16             First Canadian Place, Underground city               0   
17                               Church and Wellesley               0   

   1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0            Coffee Shop                  Park                     Bakery   
1            Coffee Shop        Clothing Store  Middle Eastern Restaurant   
2                   Café           Coffee Shop                 Restaurant   
3            Coffee Shop                Bakery               Cocktail Bar   
4            Coffee Shop    Italian Restaurant                       Café   
6            Coffee Shop                  Café                        Bar   
7            Coffee Shop                 Hotel                   Aquarium   
8            Coffee Shop                  Café                      Hotel   
9            Coffee Shop                  Café                      Hotel   
14           Coffee Shop            Restaurant                       Café   
15           Coffee Shop            Restaurant                     Market   
16           Coffee Shop                  Café                 Restaurant   
17           Coffee Shop   Japanese Restaurant           Sushi Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                    Pub               Theater    Mexican Restaurant   
1         Cosmetics Shop  Fast Food Restaurant                  Café   
2     Italian Restaurant                 Hotel        Clothing Store   
3     Seafood Restaurant        Farmers Market              Beer Bar   
4           Burger Joint        Sandwich Place        Ice Cream Shop   
6        Thai Restaurant            Steakhouse                Bakery   
7                   Café   Fried Chicken Joint        Scenic Lookout   
8                    Bar            Restaurant   American Restaurant   
9    American Restaurant            Restaurant    Seafood Restaurant   
14   Japanese Restaurant                Bakery    Italian Restaurant   
15                  Park    Italian Restaurant           Pizza Place   
16                 Hotel            Steakhouse    Seafood Restaurant   
17               Gay Bar            Restaurant             Gastropub   

   7th Most Common Venue     8th Most Common Venue 9th Most Common Venue  \
0                   Café                Restaurant        Breakfast Spot   
1                 Bakery       Japanese Restaurant       Bubble Tea Shop   
2                 Bakery            Cosmetics Shop                 Diner   
3                   Café               Cheese Shop            Steakhouse   
4                    Spa                       Bar    Chinese Restaurant   
6       Asian Restaurant                Restaurant          Burger Joint   
7             Restaurant        Italian Restaurant               Brewery   
8              Gastropub        Seafood Restaurant    Italian Restaurant   
9                 Bakery                Steakhouse                   Gym   
14    Seafood Restaurant                  Beer Bar                 Hotel   
15                Bakery        Chinese Restaurant                   Pub   
16   American Rest

### Cluster 2

In [53]:
dtoronto_merged.loc[dtoronto_merged['Cluster Labels'] == 1, dtoronto_merged.columns[[2] + list(range(5, dtoronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
13     Rosedale               1                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
13                 Trail           Yoga Studio    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
13  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
13   Dumpling Restaurant             Donut Shop

### Cluster 3

In [54]:
dtoronto_merged.loc[dtoronto_merged['Cluster Labels'] == 2, dtoronto_merged.columns[[2] + list(range(5, dtoronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
12  CN Tower, Bathurst Quay, Island airport, Harbo...               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
12       Airport Service        Airport Lounge         Boat or Ferry   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
12       Harbor / Marina              Boutique   Rental Car Location   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
12                   Bar           Coffee Shop      Sculpture Garden   

   10th Most Common Venue  
12     Airport Food Court

### Cluster 4

In [55]:
dtoronto_merged.loc[dtoronto_merged['Cluster Labels'] == 3, dtoronto_merged.columns[[2] + list(range(5, dtoronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
10             Harbord, University of Toronto               3   
11  Chinatown, Grange Park, Kensington Market               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                  Café            Restaurant    Italian Restaurant   
11                  Café                   Bar    Chinese Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
10   Japanese Restaurant             Bookstore        Sandwich Place   
11           Coffee Shop   Dumpling Restaurant    Mexican Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10                Theater                Bakery                   Bar   
11  Vietnamese Restaurant                Bakery           Record Shop   

   10th Most Common Venue  
10                    Pub  
11            Gaming Cafe

### Cluster 5

In [56]:
dtoronto_merged.loc[dtoronto_merged['Cluster Labels'] == 4, dtoronto_merged.columns[[2] + list(range(5, dtoronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5     Christie               4         Grocery Store                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5                  Park             Nightclub                 Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                  Bank           Candy Store    Italian Restaurant   

  9th Most Common Venue 10th Most Common Venue  
5            Baby Store             Restaurant

Based on the result of 5 clusters, we can determine a name to each cluster. For this clustering, I assign the names for each cluster will be : 

**Cluster 1** : Coffee Shop and Restaurant<br>
**Cluster 2** : Park and Playground <br>
**Cluster 3** : Airport and Transportation Services <br>
**Cluster 4** : Cafe and Restaurant <br>
**Cluster 5** : Common Services <br>